### ISSUE TITLE
RULE A VIOLATION: a Q4_K .apr produces garbage on CUDA on BOTH hosts while answering correctly on CPU — rc=0, no fallback, recorded GREEN

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3922]
## Measured, both required hosts, same file, same binary

```
qwen2.5-coder-1.5b-instruct-q4k.apr        a Q4_K model

CUDA, lambda:  rc=0  ran="gpu"  fell_back=false
               Assistant: "ürnópez zombie.ERRópez zombieópez zombie zombie zombie…"
CUDA, gx10:    identical text
CPU,  both:    "The capital of France is Paris."
```

**Correct on CPU. Garbage on CUDA. Both hosts. No fallback. `rc=0`.** And the ladder records the row **GREEN**, because its `chat` leg is judged by the exit code (#3921).

## This is Rule A's subject, unambiguously

Operator, 2026-09-21, verbatim with *"the end."* appended:

> **"ALL Q4_K models must work on CUDA on gx10 AND lambda before any release; 'optional rung' is no excuse."**

This is **not** the `tinyllama` shape. `tinyllama` fails identically on both backends, which is why three sessions independently concluded Rule A's subject was absent there and the row blocks on model-support grounds instead (#3914, #3900 Q3).

**Here the subject is present:** the model works on one backend and computes a different function on the other. That is PMAT-785's *"silent Q4_K-decode garbage"* — the exact defect Rule A was written to forbid, and the exact defect EPIC #3477 exists for. It was invisible because the gate read a status code.

## How it was found, because the provenance matters

1. #3911 (embedded `.apr` tokenizer) introduced a `CUDA_ERROR_ILLEGAL_ADDRESS` that made `chat` fail `rc=8`.
2. #3920 fixed that crash — **and removed the thing that was hiding this.**
3. The discriminating pair then exonerated the tokenizer completely:

```
EMBEDDED tokenizer + CPU    ->  correct answer
SIBLING  tokenizer + --gpu  ->  IDENTICAL garbage
```

**The variable is the backend, not the encoder.** Neither tokenizer path is implicated; both produce garbage on CUDA and correct output on CPU.

So this is a **pre-existing GPU defect, revealed rather than caused** — third instance tonight of one defect sitting in front of another.

## Why it is not a duplicate of the rows already open

- **#3921** is the *gate* defect — `chat`/`code` judged by `rc`, `serve` by http, output judged only by `apr qa`'s golden leg. Fixing it makes this **visible**; it does not fix it. *Fixing the gate that detects a defect does not fix the defect.*
- **#3885** (f16 `.apr` dispatched into the Q4K pool path) is a different file class — that one 500s at the serve layer and is fixed. This model is Q4_K, reaches the GPU, returns `rc=0`, and computes wrong.
- **#3908** is bf16 `.apr` declining CUDA entirely. This one does **not** decline; it runs and is wrong.

## Suggested first measurements, cheapest first

The apparatus for this exists and is warm — the IQ4_NL / IQ3_S / Q5_1 device A/B harness (#3884, #3885) compares a GPU kernel against the CPU decoder on-device with planted faults proven RED.

1. **Is the CPU/GPU divergence at the weights or at the compute?** Dequantize one `blk.0.*` tensor on both paths and compare element-wise. A layout/stride difference and an arithmetic difference look different.
2. **Is it `.apr`-specific or Q4_K-specific?** The same model as GGUF (`qwen2.5-coder-1.5b-instruct-q4_k_m.gguf`) is on both hosts and is **green**. If the GGUF is correct on CUDA and the `.apr` is not, the defect is in the `.apr` GPU upload path, not the Q4_K kernel — which would make LAYOUT-001/002 the first place to look (`contracts/tensor-layout-v1.yaml` is the source of truth, and CLAUDE.md records that this class has occurred 100+ times).
3. **Vary the prompt** — one failing input is an anecdote, and that rule has already inverted a GPU diagnosis in this release.

**Do not conclude from the text alone.** `"zombie"` repeating is structured degeneracy, which points at wrong logits rather than unreadable weights — but that inference was wrong once tonight already.

## Anti-vacuity control (required)

The control is a **real artifact on a real host**, which is unusually strong:

- the same file must produce **correct output on CPU** and **garbage on CUDA** before the fix — both arms measured, on both hosts;
- after the fix, CUDA must produce output that passes the golden patterns, and **CPU must be unchanged** (a "fix" that changes the CPU answer has altered the reference);
- the GGUF sibling, currently green on CUDA, must **stay** green — otherwise the fix has broken the path that works.

Refs #3921, #3477, #3885, #3908, #3914, PMAT-785

**Comment by noahgift** [comment]
## CORRECTION by the filer: "the variable is the backend" was NOT established, and the defect is narrower and better located

I wrote into this ticket's brief:

> *"The variable is the backend, not the encoder."*

**That was not supported, and I should not have stated it.** It came from a discriminating pair that changed **two** variables at once:

```
EMBEDDED tokenizer + CPU    ->  correct
SIBLING  tokenizer + --gpu  ->  garbage
```

Tokenizer **and** verb/model-class both move between those rows. **A discriminating pair that changes two variables discriminates nothing** — and I propagated the conclusion as established rather than checking whether the pair isolated anything.

### The matrix that does isolate it

One binary (`apr 0.69.1 (92858b7fd)`, sha256 `fe773a179eabd2b45605f2de`), one session, one RTX 4090, model sha `187234f417d5c387230f`:

```
verb  backend  model class                                     result
run   GPU      try_apr_cuda_inference / OwnedQuantizedModel    "The capital of France is Paris."
run   CPU      OwnedQuantizedModel                             "The capital of France is Paris."
chat  CPU      AprV2Model                                      "The capital of France is Paris."
chat  GPU      AprV2ModelCuda                                  "oooooooo… ustedópez zombieópez zombie…"
```

**`apr run --gpu` and `apr chat --gpu` on the same `.apr`, the same card, in the same session, give completely different answers. The variable is the MODEL CLASS.**

There are **two `.apr` GPU implementations** and only one is broken: `AprV2ModelCuda` (what `apr chat` uses). `OwnedQuantizedModel::from_apr` via `try_apr_cuda_inference` (what `apr run --gpu` uses) is correct on the same file.

### The tokenizer is exonerated — for a different reason than the brief gave

`chat` + CPU loads the **same sibling tokenizer** and answers correctly. So sibling-vs-embedded is not the variable at all.

(Recorded separately, not this bug: embedded reports **25** special tokens, sibling **22**. That is #3920's surface and worth its own row.)

### My suggested first measurement was clean AND would have misled us

The brief said to compare the `.apr` against its green GGUF sibling first, as a cheap way to split the search space. Measured: **`.apr`/GPU, `.apr`/CPU, GGUF/GPU, GGUF/CPU — all four CORRECT** via `apr run`. Then eight further cells (2 `.apr` variants × 4 prompts, all GPU) — all coherent.

**Under that test the defect is invisible, because `apr run` does not use the broken path.** The axis that mattered was **verb**, not format. Rule 6 (vary the input) was applied properly and found nothing, because the path under test was not the path that breaks.

**Worth stating as a rule: a clean measurement can be clean because it misses the defect. "I varied the input and it held" is evidence only if the path under test is the path that breaks.**

## Revised blast radius, and it narrows Rule A rather than dissolving it

The ladder's legs, measured:

```
model_ladder.sh:566   apr chat  -> AprV2ModelCuda        BROKEN, judged by rc until #3921
model_ladder.sh:530   apr run   -> OwnedQuantizedModel   correct
model_ladder.sh:458   apr qa    -> golden_output         correct — the ONLY leg that judged output
```

**The only leg that ever judged output was exercising the implementation that works.** That is the complete explanation for why these rows read green, and it is not coincidence — the leg with real judging and the leg with the defect were different legs.

So: **`.apr` × GPU × the `chat` verb.** Roughly five `.apr` files on lambda, four on gx10. **Not** a Q4_K kernel defect, **not** GGUF, **not** all four verbs.

**Rule A**, on the narrow reading (*does this Q4_K model work on CUDA*): `apr run --gpu` says **yes**. On the operator's widened four-verb bar (`run`, `chat`, `code`, `serve`): `chat` is one of the four and it does **not**. **The violation stands, as one verb on one model-class implementation** — not as silent Q4_K decode garbage in the kernel. The title overstates it and the body now does not.

## Two secondary signals from the failing cell, not yet investigated

```
0.8 tok/s  (32 tokens in 41.7 s)
"Pre-cached 934 MB of weights (197 quantized, 0 F32 tensors)"
"This APR model uses 24035 MB VRAM (F32 tensors)"  + a tip to quantize an already-q4k model
```

**Both VRAM claims cannot be true and they are in the same log.** 24 GB for a 1.1 GB q4k file is ~22×, close to what treating quantized bytes as F32 would give — which would explain the garbage, the throughput and the VRAM claim with one mechanism. **A hypothesis with a mechanism, not a diagnosis**; being measured next.

## Reproduction note

`apr chat` has **no `--prompt` flag and hangs on a pipe.** It needs a pty:

```bash
script -qec "apr chat model.apr --gpu --temperature 0 --max-tokens 32"   # prompt on stdin, sleep before exit
```

Nine minutes were lost to this. It is also arguably its own row: **a verb with no non-interactive mode is a verb no gate can drive except through a pty trick** — which is why the ladder pipes into it and gets an `rc` rather than output.

Credit: the matrix, all three corrections and the pty note are aprender-d8's.

Refs #3921, #3920, #3477, PMAT-785

**Comment by noahgift** [comment]
milestone → 0.71.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: model/backend correctness or the certified matrix: Don't Leave Behind (#3994)
